### ISSUE TITLE
ILogger interface is difficult to extend/decorate

### ISSUE THREAD

**Issue by nehmebilal** [aspnet/Logging#848]
I find it difficult to work with the `ILogger` interface, in particular the method below:

```csharp
void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter);
```
I tried to write a decorator for the ILogger interface to append a request id to every message being logged but that was not possible because it's hard (or even not possible) to recover the original logged message from the `TState state`. This is because the implementation used to create the state is in an extension method.

I believe that the `ILogger` interface should contain the method below instead:

```csharp
void Log(LogLevel logLevel, EventId eventId, Exception exception, string message, params object[] args);
```

Which makes it easy to extend the ILogger using a decorator. Below an example:

```csharp
public void Log(LogLevel logLevel, EventId eventId, Exception exception, string message, params object[] args)
{
       string requestId = requestIdAccessor.GetRequestId();
       message += ", RequestId: {RequestId}";
       args = args.Concat(new[] { requestId }).ToArray();
       logger.Log(logLevel, eventId, exception, message, args);
}
```

Note that I couldn't use scopes because my `ILogger` is shared between multiple threads that are associated with different requests. Because `BeginScope` changes the state of the logger, it's not safe to use in multiple threads (which is another issue with ILogger).

**Comment by davidfowl** [comment]
The logger interface was made this way so that complex structure could be logged (potentially without allocation or boxing). That’s why it’s generic. 

The method you suggest isn't just  breaking, it’s extremely inefficient.

Yea it wasn’t designed for that but we’re not going to break the interface. Using a provider that supports logging arbitrary state like Serilog might be a better solution.

I’d love to understand why scopes don’t work here. Can you show a minimal code sample.

**Comment by ddobric** [comment]
I agree with @davidfowl . Your suggestion wouldn't really break interface (not now at least), but it is not the best way to achieve what you want. We typically use scopes in this case.
But I also agree that TState was not done best way.

Regarding using scopes in this context, scopes are deigned to represent a graph. It is nothing bad about that. However in many use cases, when building distributed systems people rather want to 'tag' messages.

Information like machine name, user name etc. are rather represented as tags than graph information.
The best solution would be to have a possibility to use both. 

For example. If you run on a single machine, scope would be "machinename/username".
If you run in a cluster you would rather expect "username/machinename".

However we also have cases in distributed systems, when user context is switching (impersonation) across different services. This makes scope representation very hard.

To recap, yes we can mostly use or miss-use scopes. But, as shown in example above, good API design should not force us to use A instead of B if infrastructure changes or application architecture changes.

**Comment by nehmebilal** [comment]
Thanks for responses.

@davidfowl If I have a singleton `ILogger<MySingletonService>` injected into `MySingletonService`, and that `MySingletonService` is called from multiple threads where the requestId is different in each, two scopes will be potentially active at the same time resulting in both request ids being logged where I only want the right request id to go with each log. Does that make sense? 

@ddobric How would you implement tagging by request id as discussed above? (given the race condition between multiple requests).

**Comment by nehmebilal** [comment]
Interesting [article](https://www.infoq.com/articles/default-interface-methods-cs8?utm_source=infoq&utm_medium=popular_widget&utm_campaign=popular_content_list&utm_content=homepage) that discusses a similar example and proposes the new default methods in C# 8 as a potential solution.

On another note, just to highlight how inconvenient the ILogger interface is, see [here](https://github.com/Azure/diagnostics-eventflow/blob/09d582d0a1a1e0a8c7c06ca7faebdefaff8fbff5/src/Microsoft.Diagnostics.EventFlow.Inputs.MicrosoftLogging/Logger/EventFlowLogger.cs) the kind of hacking that *Azure/diagnostics-eventflow* guys needed to do in order to work with the `ILogger` interface (code snippet below). They basically needed to case the state down `FormattedLogValues` which is an implementation detail of the `ILogger` default implementation. To me, that is a clear indication that there is a lot wrong with the interface.

```csharp
        public void Log<TState>(Extensions.Logging.LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
        {
            Validation.Requires.NotNull(formatter, nameof(formatter));

            if (!IsEnabled(logLevel))
                return;

            var message = formatter(state, exception);

            Dictionary<string, object> properties = new Dictionary<string, object>();
            if (state is FormattedLogValues)
            {
                var formattedState = state as FormattedLogValues;
                //last KV is the whole message, we will pass it separately
                for (int i = 0; i < formattedState.Count - 1; i++)
                    properties.Add(formattedState[i].Key, formattedState[i].Value);
            }
```

**Comment by davidfowl** [comment]
> @davidfowl If I have a ILogger<MySingletonService> injected into a singleton MySingletonService, and that MySingletonService is called from multiple threads where the requestId is different, two scopes will be created (if I use scopes to tag by request id) resulting in both request ids being logged where I only wanted the right request id to go with the right log. Does that make sense?

Scoping should work fine here actually. Scopes are async local so as long as you're not overlapping the same "async flow" it'll be fine. Each call will get the appropriate request id.

**Comment by nehmebilal** [comment]
@davidfowl Thanks for the clarification, this is very useful information and makes scopes more useful for me. I know that we had a scenario where the requests ids were being mixed up, I believe it was inside an [Orleans](github.com/dotnet/orleans) context (I'll try to find out more info).

That said, I still think the `ILogger` interface needs work as discussed in my previous comment above.

**Comment by sergeybykov** [comment]
>@sergeybykov does this work in an Orleans context?

I need to page @xiazen and @julido.

**Comment by davidfowl** [comment]
> That said, I still think the ILogger interface needs work as discussed in my previous comment above.

Except we're not going to make any breaking changes.

**Comment by nehmebilal** [comment]
> Except we're not going to make any breaking changes.

Once [default interface methods](https://github.com/dotnet/csharplang/blob/master/proposals/default-interface-methods.md) are available in C#, you can add the higher level Log method discussed in the description without breaking anyone (it would simply delegate to the existing Log method).

**Comment by davidfowl** [comment]
> Once default interface methods are available in C#, you can add the higher level Log method discussed in the description without breaking anyone (it would simply delegate to the existing Log method).

Maybe. I wouldn't jump on that as the holy grail for changing interfaces. Right now scopes are our answer here and there's no plan to change that.

**Comment by nehmebilal** [comment]
Haha I thought I presented enough evidence that the issues with the interface are beyond what scopes (a concrete use case in event flow is loud and clear). But so far it sounds like you chose to resist feedback over acknowledging that there is a problem that should be addressed in one way or another.

I am not sure I'll be motivated to share feedback next time..

**Comment by xiazen** [comment]
@nehmebilal if your question is does scoping work in orleans, the answer is yes. Orleans is integrated with `Microsoft.Extensions.Logging`, including scoping.

**Comment by aspnet-hello** [comment]
We periodically close 'discussion' issues that have not been updated in a long period of time.

We apologize if this causes any inconvenience. We ask that if you are still encountering an issue, please log a new issue with updated information and we will investigate.
